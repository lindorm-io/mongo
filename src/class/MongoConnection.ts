import { Db, MongoClient } from "mongodb";
import { IMongoConnectionOptions } from "../typing";

export class MongoConnection {
  private options: IMongoConnectionOptions;
  private client: MongoClient;
  private database: Db;

  constructor(options: IMongoConnectionOptions) {
    this.options = options;
  }

  private getUrl(): string {
    const { user, password, host, port } = this.options;
    return `mongodb://${user}:${password}@${host}:${port}/`;
  }

  public async connect(): Promise<void> {
    this.client = await MongoClient.connect(this.getUrl(), {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
    this.database = await this.client.db(this.options.name);
  }

  public db(): Db {
    if (!this.database) {
      throw new Error("You must connect() before you can get db()");
    }
    return this.database;
  }

  public async disconnect(): Promise<void> {
    await this.client.close();
  }
}
