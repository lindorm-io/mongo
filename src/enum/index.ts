export * from "./mongo-connection-type";
