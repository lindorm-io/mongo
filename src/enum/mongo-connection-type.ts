export enum MongoConnectionType {
  MEMORY = "MEMORY",
  STORAGE = "STORAGE",
}
