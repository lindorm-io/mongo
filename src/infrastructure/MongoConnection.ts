import { IMongoConnection, IMongoConnectionOptions, TMongoDatabase } from "../typing";
import { MongoConnectionBase } from "./MongoConnectionBase";
import { MongoConnectionStorage } from "./MongoConnectionStorage";
import { MongoConnectionMemory } from "./MongoConnectionMemory";
import { MongoConnectionType } from "../enum";

export class MongoConnection implements IMongoConnection {
  private connection: MongoConnectionBase;

  constructor(options: IMongoConnectionOptions) {
    switch (options.type) {
      case MongoConnectionType.STORAGE:
        this.connection = new MongoConnectionStorage(options);
        break;

      case MongoConnectionType.MEMORY:
        this.connection = new MongoConnectionMemory(options);
        break;

      default:
        throw new Error(`Unknown connection type: [ ${options.type} ]`);
    }
  }

  public async connect(): Promise<void> {
    return this.connection.connect();
  }

  public async disconnect(): Promise<void> {
    return this.connection.disconnect();
  }

  public getDatabase(): TMongoDatabase {
    return this.connection.getDatabase();
  }
}
