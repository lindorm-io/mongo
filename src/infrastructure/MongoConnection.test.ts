import { MongoConnectionType } from "../enum";
import { MongoConnection } from "./MongoConnection";
import { TObject } from "@lindorm-io/core";

describe("MongoConnection", () => {
  let inMemoryStore: TObject<any>;
  let connection1: MongoConnection;
  let connection2: MongoConnection;

  beforeEach(async () => {
    inMemoryStore = { initialized: true };

    connection1 = new MongoConnection({
      auth: {
        user: "user",
        password: "password",
      },
      databaseName: "dbName",
      type: MongoConnectionType.MEMORY,
      url: {
        host: "host",
        port: 1234,
      },
      inMemoryStore,
    });
    connection2 = new MongoConnection({
      auth: {
        user: "user",
        password: "password",
      },
      databaseName: "dbName",
      type: MongoConnectionType.MEMORY,
      url: {
        host: "host",
        port: 1234,
      },
      inMemoryStore,
    });

    await connection1.connect();
    await connection2.connect();
  });

  test("should persist", async () => {
    const db = connection1.getDatabase();

    const collection = await db.collection("coName");
    await collection.createIndex({ index: true }, { options: true });

    await collection.insertOne({
      id: "1",
      group: "1",
      version: 0,
      data: { string: "string", number: 123, bool: true, arr: [] },
    });
    await collection.insertOne({
      id: "2",
      group: "1",
      version: 0,
      data: { string: "string", number: 123, bool: true, arr: [] },
    });
    await collection.insertOne({
      id: "3",
      group: "2",
      version: 0,
      data: { string: "string", number: 123, bool: true, arr: [] },
    });
    expect(collection).toMatchSnapshot();

    await collection.findOneAndUpdate(
      { id: "2", version: { $eq: 0 } },
      {
        $set: {
          id: "2",
          group: "1",
          version: 1,
          data: { bool: true, arr: [] },
        },
      },
    );
    expect(collection).toMatchSnapshot();

    await expect(collection.findOne({ id: "1" })).resolves.toMatchSnapshot();

    await expect(collection.find({ group: "1" })).resolves.toMatchSnapshot();

    await collection.findOneAndDelete({ id: "1" });
    expect(inMemoryStore).toMatchSnapshot();

    await collection.deleteMany({ group: "1" });
    expect(inMemoryStore).toMatchSnapshot();

    await connection1.disconnect();
    await connection2.disconnect();

    expect(connection1.getDatabase()).toMatchSnapshot();
    expect(connection2.getDatabase()).toMatchSnapshot();

    expect(inMemoryStore).toMatchSnapshot();
  });
});
