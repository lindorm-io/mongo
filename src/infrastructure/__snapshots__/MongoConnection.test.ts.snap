// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoConnection should persist 1`] = `
MongoInMemoryCollection {
  "collection": Object {
    "data": Array [
      Object {
        "data": Object {
          "arr": Array [],
          "bool": true,
          "number": 123,
          "string": "string",
        },
        "group": "1",
        "id": "1",
        "version": 0,
      },
      Object {
        "data": Object {
          "arr": Array [],
          "bool": true,
          "number": 123,
          "string": "string",
        },
        "group": "1",
        "id": "2",
        "version": 0,
      },
      Object {
        "data": Object {
          "arr": Array [],
          "bool": true,
          "number": 123,
          "string": "string",
        },
        "group": "2",
        "id": "3",
        "version": 0,
      },
    ],
    "indices": Array [
      Object {
        "index": Object {
          "index": true,
        },
        "options": Object {
          "options": true,
        },
      },
    ],
  },
}
`;

exports[`MongoConnection should persist 2`] = `
MongoInMemoryCollection {
  "collection": Object {
    "data": Array [
      Object {
        "data": Object {
          "arr": Array [],
          "bool": true,
          "number": 123,
          "string": "string",
        },
        "group": "1",
        "id": "1",
        "version": 0,
      },
      Object {
        "data": Object {
          "arr": Array [],
          "bool": true,
          "number": 123,
          "string": "string",
        },
        "group": "2",
        "id": "3",
        "version": 0,
      },
      Object {
        "data": Object {
          "arr": Array [],
          "bool": true,
        },
        "group": "1",
        "id": "2",
        "version": 1,
      },
    ],
    "indices": Array [
      Object {
        "index": Object {
          "index": true,
        },
        "options": Object {
          "options": true,
        },
      },
    ],
  },
}
`;

exports[`MongoConnection should persist 3`] = `
Object {
  "data": Object {
    "arr": Array [],
    "bool": true,
    "number": 123,
    "string": "string",
  },
  "group": "1",
  "id": "1",
  "version": 0,
}
`;

exports[`MongoConnection should persist 4`] = `
MongoInMemoryCursor {
  "data": Array [
    Object {
      "data": Object {
        "arr": Array [],
        "bool": true,
        "number": 123,
        "string": "string",
      },
      "group": "1",
      "id": "1",
      "version": 0,
    },
    Object {
      "data": Object {
        "arr": Array [],
        "bool": true,
      },
      "group": "1",
      "id": "2",
      "version": 1,
    },
  ],
}
`;

exports[`MongoConnection should persist 5`] = `
Object {
  "dbName": Object {
    "coName": Object {
      "data": Array [
        Object {
          "data": Object {
            "arr": Array [],
            "bool": true,
            "number": 123,
            "string": "string",
          },
          "group": "2",
          "id": "3",
          "version": 0,
        },
        Object {
          "data": Object {
            "arr": Array [],
            "bool": true,
          },
          "group": "1",
          "id": "2",
          "version": 1,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "index": true,
          },
          "options": Object {
            "options": true,
          },
        },
      ],
    },
  },
  "initialized": true,
}
`;

exports[`MongoConnection should persist 6`] = `
Object {
  "dbName": Object {
    "coName": Object {
      "data": Array [
        Object {
          "data": Object {
            "arr": Array [],
            "bool": true,
            "number": 123,
            "string": "string",
          },
          "group": "2",
          "id": "3",
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "index": true,
          },
          "options": Object {
            "options": true,
          },
        },
      ],
    },
  },
  "initialized": true,
}
`;

exports[`MongoConnection should persist 7`] = `
MongoInMemoryDatabase {
  "database": Object {
    "coName": Object {
      "data": Array [
        Object {
          "data": Object {
            "arr": Array [],
            "bool": true,
            "number": 123,
            "string": "string",
          },
          "group": "2",
          "id": "3",
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "index": true,
          },
          "options": Object {
            "options": true,
          },
        },
      ],
    },
  },
  "databaseName": "dbName",
}
`;

exports[`MongoConnection should persist 8`] = `
MongoInMemoryDatabase {
  "database": Object {
    "coName": Object {
      "data": Array [
        Object {
          "data": Object {
            "arr": Array [],
            "bool": true,
            "number": 123,
            "string": "string",
          },
          "group": "2",
          "id": "3",
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "index": true,
          },
          "options": Object {
            "options": true,
          },
        },
      ],
    },
  },
  "databaseName": "dbName",
}
`;

exports[`MongoConnection should persist 9`] = `
Object {
  "dbName": Object {
    "coName": Object {
      "data": Array [
        Object {
          "data": Object {
            "arr": Array [],
            "bool": true,
            "number": 123,
            "string": "string",
          },
          "group": "2",
          "id": "3",
          "version": 0,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "index": true,
          },
          "options": Object {
            "options": true,
          },
        },
      ],
    },
  },
  "initialized": true,
}
`;
